### PR TITLE
Add --shell-auto-fallback

### PR DESCRIPTION
**Pull request by passcod**
Generates shell code that hooks into the "command not found" mechanism
and attempts to run the command with npx instead of failing. The option
has an optional argument that forces generation of a particular variant,
otherwise it attempts to autodetect the shell.

As Seen On Twitter[™](https://twitter.com/passcod/status/869469928474107906)

### CODE REVIEW

**Comment by zkat** [comment]
This is lovely. Do you mind adding docs about this to `README.md`?

**Review by zkat — CHANGES_REQUESTED** [review]
Docs, please!

File: auto-fallback.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  function __fish_command_not_found_handler --on-event fish_command_not_found
+    echo "Trying with npx..."
+    npx $argv
+  end
+
+  functions --erase __fish_command_not_found_on_interactive
+end`
+
+module.exports = function autoFallback (shell) {
+  const { BASH_VERSION, SHELL, ZSH_VERSION } = process.env
```
Comment (zkat) [inline auto-fallback.js]: We're restricted to `node@4` or latest, so was can't do destructuring assignment 😭

File: index.js [hunk 2]
```diff
@@ -18,6 +19,12 @@ updateNotifier({pkg}).notify()
 main(parseArgs())
 
 function main (argv) {
+  const shell = argv['shell-auto-fallback']
+  if (shell || shell === '') {
```
Comment (zkat) [inline index.js:23]: won't this *always* be true? `parseArgs()` defaults to `''`

File: parse-args.js [hunk 3]
```diff
@@ -62,6 +62,14 @@ function parseArgs () {
     type: 'string',
     describe: 'execute string as if inside `npm run-script`'
   })
+  .option('shell-auto-fallback', {
+    choices: ['', 'bash', 'fish', 'zsh'],
```
Comment (zkat) [inline parse-args.js:66]: Why `''`?

**Review by zkat — CHANGES_REQUESTED** [review]
Ok! I actually went and reviewed the code, which I should've done in the first place. Looks good so far, with a few changes requested.

This still needs to get documented (be sure to add it to `usage` in both the README and in `parse-args`!)

File: parse-args.js (hunk 3 above)
Comment (passcod) [reply parse-args.js:66]: I'll re-test, but I think yargs refused to parse the option if the default wasn't in the choices. Although. Given the difficulty of figuring out the shell from a subprocess, maybe I should just remove both the autodetecting code and the default value here, and just do it based on the passed value.

Comment (zkat) [reply parse-args.js:66]: Just don't give it a default?

File: index.js (hunk 2 above)
Comment (passcod) [reply index.js:23]: `requireArg` is `false`! So sometimes the option will return null/undefined/false/whatever yargs returns for a non-provided option. But `''` is falsy, so I do need to check.

File: parse-args.js (hunk 3 above)
Comment (passcod) [reply parse-args.js:66]: As to _why I made the default `''`_... err. I don't remember. I'll have a look at the code again in its entirety to see if I can figure it out.

File: index.js (hunk 2 above)
Comment (zkat) [reply index.js:23]: doesn't having a default mean that you always get `''` if you don't provide the option, though? I'm so confused.

**Comment by passcod** [comment]
Oops, I broke it. ...good reminder to add tests, I suppose. I'll do that now.
Orrrr maybe it never worked. I've confused myself. I'll figure it out.

**Comment by passcod** [comment]
Ok, I've fixed it, added tests, wrote documentation. I've kept the same functionality, but I can remove the autodetect / default if you want.

**Review by zkat — APPROVED** [review]

**Pull request merged**